Support tron StakeV2 transactions

ClarkChenc opened the pull request:
StakeV2.0 add new types of transactions: 
* FreezeBalanceV2
* UnFreezeBalanceV2
* DelegateResource
* UndelegateResource
* WithdrawExpireUnfreeze

This pr supports these new transactions above.

Here are operation examples related with these transactions:
1. FreezeBalanceV2
<img width="497" alt="image" src="https://user-images.githubusercontent.com/9441239/220323960-9454f701-67fa-4268-ae3f-9f2cb92a7c28.png">
<img width="914" alt="image" src="https://user-images.githubusercontent.com/9441239/220517302-d51aa245-4dcf-45f7-a925-98aa4ba96c03.png">
<img width="860" alt="image" src="https://user-images.githubusercontent.com/9441239/220517340-7b6eb8b0-4e29-4a4a-b999-2d971cae638c.png">
<img width="867" alt="image" src="https://user-images.githubusercontent.com/9441239/220517369-cf9d8476-bf03-4996-a7da-cc2131184180.png">
<img width="925" alt="image" src="https://user-images.githubusercontent.com/9441239/220517399-05b14a72-855a-40cf-b1c2-6b408ede1e23.png">
<img width="768" alt="image" src="https://user-images.githubusercontent.com/9441239/220517430-9c89ad24-e436-4645-8bb0-036721725f35.png">
<img width="872" alt="image" src="https://user-images.githubusercontent.com/9441239/220517476-b9223ed2-1551-4bed-8f41-3eca7da51422.png">
<img width="495" alt="image" src="https://user-images.githubusercontent.com/9441239/220324430-1beed736-81d8-4d27-9724-eac9af9a8361.png">

2. UnFreezeBalanceV2
<img width="493" alt="image" src="https://user-images.githubusercontent.com/9441239/220327583-1688ad8b-359b-43aa-925c-31345d829f78.png">
<img width="1030" alt="image" src="https://user-images.githubusercontent.com/9441239/220516981-49e99d8e-d5b6-4d5e-8d97-28e652413c18.png">
<img width="883" alt="image" src="https://user-images.githubusercontent.com/9441239/220517020-4350dd44-891e-4c04-ab0f-088c5a9d414d.png">
<img width="862" alt="image" src="https://user-images.githubusercontent.com/9441239/220517058-30491c91-96e5-471c-88d8-8ee9f0aa9fd7.png">
<img width="899" alt="image" src="https://user-images.githubusercontent.com/9441239/220517093-b1d658f5-1352-47b3-af6b-512b11b768cf.png">
<img width="875" alt="image" src="https://user-images.githubusercontent.com/9441239/220517122-f9ddbd80-fbb3-4bff-9da0-e7616cd16ebb.png">
<img width="849" alt="image" src="https://user-images.githubusercontent.com/9441239/220517155-cb199527-5641-425d-80bc-aa02471777fd.png">
<img width="488" alt="image" src="https://user-images.githubusercontent.com/9441239/220328216-e7525d5a-621c-4777-b475-c5139ce0f2ae.png">

3. DelegateResource
<img width="488" alt="image" src="https://user-images.githubusercontent.com/9441239/220329083-7a5e9068-9302-4ddc-a30f-510393e3c36c.png">
<img width="906" alt="image" src="https://user-images.githubusercontent.com/9441239/220329637-5a1f869e-64f9-486c-b87b-2d6871ee689f.png">
<img width="896" alt="image" src="https://user-images.githubusercontent.com/9441239/220329744-52d161eb-9ce4-4799-9564-6813944d15b6.png">
<img width="963" alt="image" src="https://user-images.githubusercontent.com/9441239/220329784-cd830587-1ece-4405-880c-c0c422deee0a.png">
<img width="995" alt="image" src="https://user-images.githubusercontent.com/9441239/220329857-b14d6d08-6394-439e-9655-26021d890fab.png">
<img width="962" alt="image" src="https://user-images.githubusercontent.com/9441239/220329932-1b852dc1-5fe5-4b1f-9e2c-acbbe8f0be41.png">
<img width="990" alt="image" src="https://user-images.githubusercontent.com/9441239/220329976-56d484a9-8c41-4d74-bacb-42556e708e86.png">
<img width="986" alt="image" src="https://user-images.githubusercontent.com/9441239/220330106-912f0656-8674-4c9f-a620-24e2527e0567.png">
<img width="491" alt="image" src="https://user-images.githubusercontent.com/9441239/220329281-8af1296d-a21b-42f3-9dcb-115be0d58f8d.png">

4. UndelegateResource
<img width="488" alt="image" src="https://user-images.githubusercontent.com/9441239/220330452-6dcfaaf4-07bf-450b-b179-8b15b7ef64fc.png">
<img width="918" alt="image" src="https://user-images.githubusercontent.com/9441239/220331310-039224c1-7f99-41dc-94d8-dcb733e751bf.png">
<img width="1011" alt="image" src="https://user-images.githubusercontent.com/9441239/220331399-ed7de428-4b8a-4289-a00e-86cb7f5fbd22.png">
<img width="908" alt="image" src="https://user-images.githubusercontent.com/9441239/220331457-c94d1447-7f38-4aa1-bd0e-dc1076191a43.png">
<img width="912" alt="image" src="https://user-images.githubusercontent.com/9441239/220331566-ab9b1730-8a15-42b5-bf83-ca6eaae552a2.png">
<img width="906" alt="image" src="https://user-images.githubusercontent.com/9441239/220331650-3747b666-2a8b-4ea8-8b2a-08c36600e260.png">
<img width="929" alt="image" src="https://user-images.githubusercontent.com/9441239/220331783-845943f4-07fd-4b64-ad03-08503c70b97c.png">
<img width="489" alt="image" src="https://user-images.githubusercontent.com/9441239/220330637-94cede85-ed54-4c77-ba83-b0ebdc8b472b.png">

5. WithdrawExpireUnfreeze
<img width="761" alt="image" src="https://user-images.githubusercontent.com/9441239/220332708-921f2f61-e0cf-421b-a3ef-06fb34982f26.png">
<img width="834" alt="image" src="https://user-images.githubusercontent.com/9441239/220332803-d0a754df-028c-4418-a9d7-b5142f26bd04.png">

More examples can also be found in examples/signStakeV2.py.

agrojean-ledger commented:
@ClarkChenc hello,

Can you rebase your branch on `develop` (and edit your PR to merge on `develop`, not master) ? 

The app's code has been updated to support Stax and some refactoring has been done. You will have to :

- put your UI code in the `ui` directory (https://github.com/LedgerHQ/app-tron/tree/develop/src/ui) in [ui_review_menu_bagl.c](https://github.com/LedgerHQ/app-tron/blob/develop/src/ui/ui_review_menu_bagl.c)
- remove any blue specific stuff (everything blue has been removed).
- add the TX review preparation in the new Stax UI code [here](https://github.com/LedgerHQ/app-tron/blob/feda295aa27f6f55711c2abc900360721421c2bc/src/ui/ui_review_menu_nbgl.c#L147)
- adapt your example code to add functional tests [here](https://github.com/LedgerHQ/app-tron/blob/feda295aa27f6f55711c2abc900360721421c2bc/tests/test_trx.py)

ClarkChenc commented:
hi，@agrojean-ledger, 

I rebased my PR on the develop branch, and modified the target merge branch to develop including adding functional tests. Looking forward to your reply if there exists any other problems.

agrojean-ledger commented on tests/test_trx.py:
```diff
@@ -590,3 +590,66 @@ def test_trx_unknown_trc20_send(self, backend, configuration, firmware,
+                lock=0))
+        self.sign_and_validate(client, firmware, 0, tx)
+
+    def test_trx_undelegate_resource(self, backend, configuration, firmware,
```
@ClarkChenc you will have to execute the tests locally with : 

```
pip install --extra-index-url https://test.pypi.org/simple/ -r requirements.txt
cd tests
pytest -v --device all --golden_run
```
And then push the new snapshots generated in `tests/snapshots`

ClarkChenc replied:
Snaphots have been added. I find ledger-secure-sdk is not easy to use, especially stax. master branch doesn't always work.

agrojean-ledger commented on src/main.c:
```diff
@@ -602,6 +602,75 @@ void handleSign(uint8_t p1,
+            else
+                strcpy(fullContract, "Energy");
+
+            print_amount(txContent.amount[0], (char *) G_io_apdu_buffer, 100, SUN_DIG);
```
@ClarkChenc Will have to run clang-format on your files with : 

 `clang-format-11 -i $(find src -type f \( -name '*.c' -o -name '*.h' \))`

so that they are compliant with the lint rules of the CI.

ClarkChenc replied:
files have been formatted, thanks a lot.

agrojean-ledger commented:
@ClarkChenc once I merge https://github.com/LedgerHQ/app-tron/pull/45 you will have to do another rebase on develop... The CI had to be updated because of some changes in our in-house reusable actions workflows.

ClarkChenc commented on src/ui/ui_review_menu_nbgl.c:
```diff
@@ -320,6 +320,72 @@ static void prepareTxInfos(ui_approval_state_t state, bool data_warning) {
             infoLongPress.text = "Confirm Approve";
             break;
+        case APPROVAL_FREEZEASSETV2_TRANSACTION:
+            txInfos.confirmCb = (nbgl_callback_t) ui_callback_address_ok;
```
yes, it's my mistake, I have fixed it.

ClarkChenc commented on tests/test_trx.py:
```diff
@@ -590,3 +590,66 @@ def test_trx_unknown_trc20_send(self, backend, configuration, firmware,
+                lock=0))
+        self.sign_and_validate(client, firmware, 0, tx)
+
+    def test_trx_undelegate_resource(self, backend, configuration, firmware,
```
tests have been formatted, thanks a lot.

ClarkChenc commented:
@agrojean-ledger
thanks to your suggestions, I have adjusted the pr to the newest ci pipeline, and added corresponding snapshots.
I wonder when will the develop branch be merged to the master branch? Is there any plan for it? thanks.

agrojean-ledger commented:
> @agrojean-ledger thanks to your suggestions, I have adjusted the pr to the newest ci pipeline, and added corresponding snapshots. I wonder when will the develop branch be merged to the master branch? Is there any plan for it? thanks.

@ClarkChenc all is fine except you need to bump the version in the `VERSION` file.

Regarding the develop branch being merged to master, it will depend on our in-house product team review of the changes in the app's code base. Can't give you a precise ETA for that.

ClarkChenc commented:
> > @agrojean-ledger thanks to your suggestions, I have adjusted the pr to the newest ci pipeline, and added corresponding snapshots. I wonder when will the develop branch be merged to the master branch? Is there any plan for it? thanks.
> 
> @ClarkChenc all is fine except you need to bump the version in the `VERSION` file.
> 
> Regarding the develop branch being merged to master, it will depend on our in-house product team review of the changes in the app's code base. Can't give you a precise ETA for that.

@agrojean-ledger version has been bummped. Thanks a lot.

agrojean-ledger commented:
CI checks are OK except static analysis because of a recent change in the reusable workflow. No big deal, I'll fix it later. 

Otherwise PR looks good : 

* new functional tests are implemented and are ok. 
* version is bumped.
* merge to develop for P4 deployment.

Merging.
